ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -393,7 +393,56 @@ public CompletableFuture<Boolean> deleteBlobAsync(BlobId blobId, long deletionTi
   @Override
   public short undeleteBlob(BlobId blobId, short lifeVersion, CloudUpdateValidator cloudUpdateValidator)
       throws CloudStorageException {
-    return 0;
+    Timer.Context storageTimer = azureMetrics.blobUndeleteLatency.time();
+    AzureBlobLayoutStrategy.BlobLayout blobLayout = azureBlobLayoutStrategy.getDataBlobLayout(blobId);
+    String blobIdStr = blobLayout.blobFilePath;
+    BlobProperties blobProperties = getBlobProperties(blobLayout);
+    Map<String, String> cloudMetadata = blobProperties.getMetadata();
+    // lifeVersion must always be present
+    short cloudlifeVersion = Short.parseShort(cloudMetadata.get(CloudBlobMetadata.FIELD_LIFE_VERSION));
+    // Below is the correct behavior. For ref, look at BlobStore::undelete and ReplicaThread::applyUndelete
+    if (cloudlifeVersion == lifeVersion && cloudMetadata.containsKey(CloudBlobMetadata.FIELD_DELETION_TIME)
+        && Long.parseLong(cloudMetadata.get(CloudBlobMetadata.FIELD_DELETION_TIME)) == Utils.Infinite_Time) {
+      String error = String.format("Cannot undelete id %s as it is undeleted in cloud", blobIdStr);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.ID_Undeleted));
+    }
+    // Don't rely on the CloudBlobStore.recentCache to do the "right" thing.
+    // Below is the correct behavior. For ref, look at BlobStore::undelete and ReplicaThread::applyUndelete
+    if (cloudlifeVersion > lifeVersion) {
+      String error = String.format("Cannot undelete id %s as it has a higher lifeVersion than the message info: %s > %s",
+          blobIdStr, cloudlifeVersion, lifeVersion);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.Life_Version_Conflict));
```

Comment:
Same here. Would returning success be correct here since this can happen when a VCR tries to replicate from a lagging replica?

Reply:
ReplicaThread does all these checks and we should almost never hit this . Throwing an error then makes more sense and it is caught in applyUndelete(). Please see ReplicaThread::applyUndelete